framework/src/main/java/org/checkerframework/framework/util/typeinference8/util/Resolution.java

```diff
@@ -215,28 +215,38 @@ private BoundSet resolveSmallestSet(Set<Variable> as, BoundSet boundSet) {
   /**
    * Resolves all the non-capture variables in {@code variables}.
    *
-   * @param variables the variables
+   * @param vars the variables
    */
-  private void resolveNoCapturesFirst(List<Variable> variables) {
+  private void resolveNoCapturesFirst(List<Variable> vars) {
     Variable smallV;
+    List<Variable> variables = new ArrayList<>(vars.size());
+    vars.forEach(
```

Comment:
`forEach` is a useful method when a lambda already exists, but I find it harder to read than a loop if the lambda has to be created.
Here, however, I think it would be clearest to use `retainAll` (with a lambda!).  Are you avoiding that for efficiency reasons?  The implementation of retainAll in ArrayList never moves any item more than once, so it is relatively efficient.

Reply:
Do you mean `removeIf`?  I don't see a version of `retainAll` that takes a function.  I will change it to:
```
    List<Variable> variables = new ArrayList<>(vars);
    variables.removeIf(v -> v.isCaptureVariable() || v.getBounds().hasInstantiation());
```